graphql-server/pkg/queries/proposal.go

```diff
@@ -25,30 +25,64 @@ type ProposalQuery struct {
 	session *bun.DB
 
 	scopedProposalStatus models.ProposalStatus
-	scopedRelatedAddress string
+	scopedAddressFilter  *models.ProposalAddressFilter
 }
 
 func NewProposalQuery(ctx context.Context, config config.Config, session *bun.DB) IProposalQuery {
 	return &ProposalQuery{ctx: ctx, config: config, session: session}
 }
 
+// generate new select query with filters applied.
+// when both status and address filters are applied, results satisfying both filters are returned.
+// when multiple sub-filters in address filter (ie. isDepositor, IsSubmitter, IsVoter), results satisfying
+// any one of the sub-filters are returned
 func (q *ProposalQuery) NewQuery() *bun.SelectQuery {
 	query := q.session.NewSelect().Model((*models.Proposal)(nil))
 
 	if q.scopedProposalStatus != "" {
 		query = query.Where("status = ?", q.scopedProposalStatus)
 	}
 
-	if q.scopedRelatedAddress != "" {
-
-		relatedDeposits := q.session.NewSelect().Model((*models.ProposalDeposit)(nil)).Column("proposal_id").Where("depositor_address = ?", q.scopedRelatedAddress)
-		relatedVotes := q.session.NewSelect().Model((*models.ProposalVote)(nil)).Column("proposal_id").Where("voter_address = ?", q.scopedRelatedAddress)
-
-		relatedProposals := relatedDeposits.Union(relatedVotes)
-
-		query = query.
-			WhereOr("id IN (?)", relatedProposals).
-			WhereOr("proposal.proposer_address = ?", q.scopedRelatedAddress)
+	if q.scopedAddressFilter != nil {
+		relatedDeposits := q.session.NewSelect().
+			Model((*models.ProposalDeposit)(nil)).
+			Column("proposal_id").
+			Where("depositor_address = ?", q.scopedAddressFilter.Address)
+
+		relatedVotes := q.session.NewSelect().
+			Model((*models.ProposalVote)(nil)).
+			Column("proposal_id").
+			Where("voter_address = ?", q.scopedAddressFilter.Address)
+
+		isFirstAddressFilter := true
+		if q.scopedAddressFilter.IsDepositor {
+			if isFirstAddressFilter {
+				query = query.Where("id IN (?)", relatedDeposits)
+				isFirstAddressFilter = false
+			} else {
+				query = query.WhereOr("id IN (?)", relatedDeposits)
+			}
+		}
+		if q.scopedAddressFilter.IsVoter {
+			if isFirstAddressFilter {
+				query = query.Where("id IN (?)", relatedVotes)
+				isFirstAddressFilter = false
+			} else {
+				query = query.WhereOr("id IN (?)", relatedVotes)
+			}
+		}
+		if q.scopedAddressFilter.IsSubmitter {
+			if isFirstAddressFilter {
+				query = query.Where("proposal.proposer_address = ?", q.scopedAddressFilter.Address)
+				isFirstAddressFilter = false
+			} else {
+				query = query.WhereOr("proposal.proposer_address = ?", q.scopedAddressFilter.Address)
+			}
+		}
+		if isFirstAddressFilter {
+			// none of the is* field are marked as true
+			query = query.Where("id IN (null)")
+		}
```

Comment:
looks weird, the idea of having a particular filter configuration that only returns nothing sounds a bit off to me :sosad:

@chihimng @cychiuae any opinion from db high hands?

Reply:
what was the original purpose?

Reply:
To achieve the behavior where
- no address filter applied -> return unfiltered results
- address filter applied but all is* fields are false -> then no results returned

Reply:
as discussed offline with @chihimng, this will be handled at the resolver level